### ISSUE TITLE
Proposal: Add Interface inheritance for definitions

### ISSUE THREAD

**Issue by SerafimArts** [facebook/graphql#335]
Spec (http://facebook.github.io/graphql/#sec-Interfaces) does not proposes this feature. But this is will be good for **ISP** (Interface Segregation Principle) as example. 

### Base definitions
```ts
interface Creatable {
    createdAt: String!
}

interface Updatable {
    updatedAt: String
}
```

### Variation of implementation

```ts
interface Person extends Creatable, Updatable {
    id: ID!
    name: String!
    # ...
}

class User implements Person {
    # Impl of interface fields: 'id', 'name', 'createdAt', 'updatedAt'
}
```

### Interface fields overriding

Discussion needed. I think. e.g. how to consider the **LSP** (Liskov Substitution Principle)?

1. Allows NonNull for Nullable fields?
2. Allows ListType for "Single" type def?
3. Allow to increase/decrease the number of arguments for field input?

```ts
# Example
interface HasDates extends Creatable, Updatable {
    updatedAt: String! # Overriding nullable field
}
```

### Errors

1. Error while cyclic inheritance

```ts
interface A extends B { ... }
interface B extends A { ... }
```

2. Error inheritance twice

```ts
interface Person extends Creatable, Updatable { ... }

type User implements Person, Creatable {
    # Error: Can not implement "Creatable" interface. Interface already defined in Person.
}
```

Like everything foreseen, there are ideas? =)

**Comment by SerafimArts** [comment]
Ooops... It seems like a duplication of https://github.com/facebook/graphql/issues/295

**Comment by tylercrompton** [comment]
It's similar but not related. Inheritance implies that the parent interface's fields are implicitly included in the child interface's schema. Additionally, you've touched on the topic of multiple inheritance and overriding properties of inherited fields. Neither of these are proposed in #295. I think that it's fair to say that this is different enough to not be a duplicate. However, the proposals conflict and thus are mutually exclusive.

**Comment by SerafimArts** [comment]
Convincingly, I'll reopen ann issue =)

**Comment by tylercrompton** [comment]
I think that for the sake of development and transition, we should defer multiple inheritance and possibly "strictification" of types as well to a separate issue, pending the adoption of the simplified version of your proposal. In other words, I think that this proposal should focus solely on regular inheritance (for now).

**Comment by Recio** [comment]
So according to this issue, it seems that implicit inheritance is still not supported? 

Is it strictly necessary to declare all the fields when implementing an interface? 
That seems to lead to a lot of duplicated code...

There is no extension / polymorphic mechanism with implicit field inheritance in graph?

**Comment by SerafimArts** [comment]
@Recio You can use custom directives to implement this behavior. This is exactly what I did:
```graphql
directive @extends(type: String!) on OBJECT

type Some { ... }
type Some2 @extends(type: "Some") { ... }
```

**Comment by Recio** [comment]
@SerafimArts I'm a bit of a newb on the matter, but I suppose on build you inject the code from Some into Some2

**Comment by SerafimArts** [comment]
@Recio in my case everything is different. I do not know how it works in the JS, because I implement my own SDL compiler =)

**Comment by leebyron** [comment]
See: https://github.com/facebook/graphql/issues/500#issuecomment-426146328

TL;DR this proposal doesn't allow new use cases that were not already possible and degrades readability of schema in the name of writing less, which goes against GraphQL's guiding principles
